<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="provoc">
<title>Lineage Definitions in ProVoC • provoc</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Lineage Definitions in ProVoC">
<meta property="og:description" content="provoc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">provoc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Package_Overview.html">ProVoC Mission Statement</a>
    <a class="dropdown-item" href="../articles/mutation_definitions.html">Lineage Definitions in ProVoC</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Lineage Definitions in ProVoC</h1>
                        <h4 data-toc-skip class="author">Devan
Becker</h4>
            
      
      
      <div class="d-none name"><code>mutation_definitions.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>This vignette demonstrates some different ways of defining and
comparing lineage definitions, including a discussion of why it
matters.</p>
</div>
<div class="section level3">
<h3 id="data-prep">Data Prep<a class="anchor" aria-label="anchor" href="#data-prep"></a>
</h3>
<p>This package contains data from <span class="citation">Baaijens et
al. (2022)</span>, as downloaded from NCBI SRA and processed via a
custom pipeline (heavily based on GromStole, with some
post-processing).</p>
<p>In this paper, they explicitly look for B.1.1.7, B.1.427, B.1.429,
and B.1.526. I’m going to add B.1.617.2 to this list, and take this list
as given (another vignette will be written to test this assumption).</p>
<p>To make the code run faster, I’ll look at every third day of
data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#library(provoc)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/load_all.html" class="external-link">load_all</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Loading <span style="color: #0000BB;">provoc</span></span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lubridate.tidyverse.org" class="external-link">lubridate</a></span><span class="op">)</span> <span class="co"># ymd()</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'lubridate'</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     date, intersect, setdiff, union</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span> <span class="co"># all paths relative to .git folder</span></span></code></pre></div>
<pre><code><span><span class="co">## here() starts at /Users/dbecker/git/DASL/provoc</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Baaijens</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Baaijens</span><span class="op">$</span><span class="va">date</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd.html" class="external-link">ymd</a></span><span class="op">(</span><span class="va">Baaijens</span><span class="op">$</span><span class="va">date</span><span class="op">)</span></span>
<span><span class="va">dates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">Baaijens</span><span class="op">$</span><span class="va">date</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dates3</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dates</span><span class="op">)</span> <span class="op">/</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">3</span></span>
<span><span class="va">dates</span> <span class="op">&lt;-</span> <span class="va">dates</span><span class="op">[</span><span class="va">dates3</span><span class="op">]</span></span>
<span></span>
<span><span class="va">b3</span> <span class="op">&lt;-</span> <span class="va">Baaijens</span><span class="op">[</span><span class="va">Baaijens</span><span class="op">$</span><span class="va">date</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">dates</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"count"</span>, <span class="st">"coverage"</span>, <span class="st">"sra"</span>, <span class="st">"date"</span>, <span class="st">"label"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">b3</span><span class="op">$</span><span class="va">mutation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parse_mutations.html">parse_mutations</a></span><span class="op">(</span><span class="va">b3</span><span class="op">$</span><span class="va">label</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">b3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     count coverage         sra       date   label        mutation</span></span>
<span><span class="co">## 774  6404    16595 SRR15505103 2021-01-17  ~5700A aa:orf1a:A1812D</span></span>
<span><span class="co">## 775 11407    12071 SRR15505103 2021-01-17 ~23401T      aa:S:Q613H</span></span>
<span><span class="co">## 776 12519    12661 SRR15505103 2021-01-17 ~23403G      aa:S:D614G</span></span>
<span><span class="co">## 777 20293    20794 SRR15505103 2021-01-17 ~14408T  aa:orf1b:P314L</span></span>
<span><span class="co">## 778  1885     3207 SRR15505103 2021-01-17 ~16500C aa:orf1b:Q1011H</span></span>
<span><span class="co">## 779  4233     4265 SRR15505103 2021-01-17  ~3037T          C3037T</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="built-in-constellation-files">Built-in Constellation Files<a class="anchor" aria-label="anchor" href="#built-in-constellation-files"></a>
</h3>
<p>The PANGO team used to maintain a GitHub repository called “<a href="https://github.com/cov-lineages/constellations" class="external-link">Constellations</a>”
<span class="citation">(<span>“Cov-Lineages/Constellations”</span>
2023)</span>, which contained mutations that were believed to be
more-or-less unique to the lineages. This is no longer updated, but is
useful for historical analyses.</p>
<p>This set of lineage definitions is available via the
<code><a href="../reference/astronomize.html">astronomize()</a></code> function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">astro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/astronomize.html">astronomize</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filter_lineages.html">filter_lineages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B.1.1.7"</span>, <span class="st">"B.1.427"</span>, <span class="st">"B.1.429"</span>, <span class="st">"B.1.526"</span>, <span class="st">"B.1.617.2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">astro</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  5 69</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">astro</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##           aa:S:L452R aa:orf3a:Q57H aa:orf1a:T265I C3037T aa:S:D614G</span></span>
<span><span class="co">## B.1.526            0             1              1      1          1</span></span>
<span><span class="co">## B.1.1.7            0             0              0      0          0</span></span>
<span><span class="co">## B.1.429            1             1              1      1          1</span></span>
<span><span class="co">## B.1.617.2          1             0              0      0          0</span></span>
<span><span class="co">## B.1.427            1             1              1      1          1</span></span></code></pre>
<p>In this matrix, the rows represent the lineages and the columns
represent the mutations, with a 1 representing the presence of a
mutation in a given lineage.</p>
<p>Notice the use of <code><a href="../reference/filter_lineages.html">filter_lineages()</a></code>. This function gets
the rows corresponding to the lineages, but it also ensures that we are
only dealing with columns that have at least one “1”. In other words, it
only uses mutations that are present in at least one lineage.</p>
<p>I have written a function to visualize this matrix:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_lineage_defs.html">plot_lineage_defs</a></span><span class="op">(</span><span class="va">astro</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/astro-defs-1.png" width="700"></p>
<p>Only the shared mutations are shown, with the last column indicating
how many unique mutations remain in each definition. Every other
mutation gets a grey background to make it easier to follow a mutation
when the matrices are larger.</p>
</div>
<div class="section level3">
<h3 id="usher-barcodes">Usher Barcodes<a class="anchor" aria-label="anchor" href="#usher-barcodes"></a>
</h3>
<p>The <a href="https://github.com/andersen-lab/Freyja" class="external-link">Freyja</a> <span class="citation">(Karthikeyan et al. 2022;
<span>“Andersen-Lab/<span>Freyja</span>”</span> 2024)</span> repository
keeps an up-to-date barcodes file, which serves the same purpose as the
definitions in constellations.</p>
<p>The <code><a href="../reference/usher_barcodes.html">usher_barcodes()</a></code> function will check for this file,
then download it if it’s not present in a number of common locations.
It’s quite a large file, so I recommend storing the file and using the
<code><a href="../reference/filter_lineages.html">filter_lineages()</a></code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bar</span> <span class="op">&lt;-</span> <span class="fu">provoc</span><span class="fu">::</span><span class="fu"><a href="../reference/usher_barcodes.html">usher_barcodes</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"working"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/filter_lineages.html">filter_lineages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B.1.1.7"</span>, <span class="st">"B.1.427"</span>, <span class="st">"B.1.429"</span>, <span class="st">"B.1.526"</span>, <span class="st">"B.1.617.2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">bar</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  5 62</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="constellations-versus-barcodes">Constellations versus Barcodes<a class="anchor" aria-label="anchor" href="#constellations-versus-barcodes"></a>
</h3>
<p>Let’s quickly look at the mutations present in <code>astro</code> but
not <code>bar</code>, then the mutations present in <code>bar</code> but
not <code>astro</code>, as well as mutations that are in both.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_lineage_defs2.html">plot_lineage_defs2</a></span><span class="op">(</span><span class="va">astro</span>, <span class="va">bar</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Constellations (red) versus Usher Barcodes (blue)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/def-compare-setdiff-1.png" width="700"></p>
<p>From this plot, we can see that there are quite a few mutations
defined in <code>astro</code> but not in <code>bar</code>. The blue bars
indicate that the mutation is used in the lineage definition in
<code>bar</code> but not <code>astro</code>; there are four mutations
defined in both B.1.1.7 and B.1.617.2 by usher barcodes that are not
present in the constellations file, and two of those mutations were also
present in B.1.429. The barcodes seem to be a little bit more lenient
about what mutations to include; constellations tried to get
more-or-less “unique” mutations.</p>
</div>
<div class="section level3">
<h3 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h3>
<p>Since we chose lineages when choosing the lineage definitions, we can
use the convenient “~ .” notation to fit the models with all
lineages.</p>
<p>Note that the Baaijens data were processed in such a way that each
sample contains the same mutations. This means that each bar you see
below is based on the exact same set of mutations. This isn’t necessary,
but I believe it improves the interpretability of the results.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span> <span class="co"># Patching ggplots together</span></span>
<span><span class="va">res_astro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/provoc.html">provoc</a></span><span class="op">(</span><span class="va">count</span> <span class="op">/</span> <span class="va">coverage</span> <span class="op">~</span> <span class="va">.</span>,</span>
<span>    lineage_defs <span class="op">=</span> <span class="va">astro</span>,</span>
<span>    data <span class="op">=</span> <span class="va">b3</span>, by <span class="op">=</span> <span class="st">"sra"</span><span class="op">)</span></span>
<span><span class="va">res_astro</span><span class="op">$</span><span class="va">date</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd.html" class="external-link">ymd</a></span><span class="op">(</span><span class="va">res_astro</span><span class="op">$</span><span class="va">date</span><span class="op">)</span></span>
<span><span class="va">res_bar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/provoc.html">provoc</a></span><span class="op">(</span><span class="va">count</span> <span class="op">/</span> <span class="va">coverage</span> <span class="op">~</span> <span class="va">.</span>,</span>
<span>    lineage_defs <span class="op">=</span> <span class="va">bar</span>,</span>
<span>    data <span class="op">=</span> <span class="va">b3</span>, by <span class="op">=</span> <span class="st">"sra"</span><span class="op">)</span></span>
<span><span class="va">res_bar</span><span class="op">$</span><span class="va">date</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd.html" class="external-link">ymd</a></span><span class="op">(</span><span class="va">res_bar</span><span class="op">$</span><span class="va">date</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gg_astro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">res_astro</span>, date_col <span class="op">=</span> <span class="st">"date"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Constellations"</span><span class="op">)</span></span>
<span><span class="va">gg_bar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">res_bar</span>, date_col <span class="op">=</span> <span class="st">"date"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Usher Barcodes"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gg_astro</span> <span class="op">/</span> <span class="va">gg_bar</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">lineage</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/gg-lapis-bar-1.png" width="700"></p>
<p>Those are two <em>very</em> different looking plots! We saw before
that the lineage definitions are different, but let’s check how they
were actually used in the data. The same function can be used from
before, which extracts the lineage definitions that were actually used
(ignoring any mutations that were part of the definition but not present
in the data).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_lineage_defs2.html">plot_lineage_defs2</a></span><span class="op">(</span><span class="va">res_astro</span>, <span class="va">res_bar</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Constellations (red) versus Usher Barcodes (blue)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/plot_lineage_defs2_with_data-1.png" width="700"></p>
<p>From this result, we can see that constellations only used four
mutations to define B.1.617.2! The barcodes used 6! This is because
there were only 4 (or 6) mutations in B.1.617.2 that were also present
in the data. Let’s dig a little deeper into this.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_actual_defs.html">plot_actual_defs</a></span><span class="op">(</span><span class="va">res_astro</span>, main <span class="op">=</span> <span class="st">"Constellations"</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/plot-actual-defs-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_actual_defs.html">plot_actual_defs</a></span><span class="op">(</span><span class="va">res_bar</span>, main <span class="op">=</span> <span class="st">"Usher Barcodes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/plot-actual-defs-2.png" width="700"></p>
<p>The two different matrices have very different definitions of
B.1.617.2, and this explains the difference in results.</p>
<p>Furthermore, B.1.427 and B.1.429 had important differences - the
barcodes have more mutations that overlap, meaning that the “de-mixing”
algorithm must choose between one or the other. I’ll expand on this in a
“Choosing Lineages to Search For” vignette.</p>
</div>
<div class="section level3">
<h3 id="lapis-cov-spectrum">LAPIS (CoV-Spectrum)<a class="anchor" aria-label="anchor" href="#lapis-cov-spectrum"></a>
</h3>
<p>Since I have not managed to find where to get an API key, the
mutations within each lineage were manually downloaded from
cov-spectrum.org <span class="citation">(Chen et al. 2022)</span>. I set
the date range to go from 2020-01-01 to 2021-04-27, with the latter
being the last sample in the Baaijens data.</p>
<p>The following code is very very hopefully going to change very soon.
An API key would be nice, and I desperately need better mutation
processing. It is, however, comforting that the column names match with
<code>bar</code>, so at least my parsing is consistent!</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lineages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B.1.1.7"</span>, <span class="st">"B.1.427"</span>, <span class="st">"B.1.429"</span>, <span class="st">"B.1.526"</span>, <span class="st">"B.1.617.2"</span><span class="op">)</span></span>
<span><span class="va">lapis_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">lineages</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"working"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">".csv"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">file</span><span class="op">$</span><span class="va">mutation</span><span class="op">[</span><span class="va">file</span><span class="op">$</span><span class="va">proportion</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lapis_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">lineages</span></span>
<span><span class="va">lapis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lineage_defs_from_list.html">lineage_defs_from_list</a></span><span class="op">(</span><span class="va">lapis_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Avoiding parsing deletions, only looking at substitutions</span></span>
<span><span class="va">lapis</span> <span class="op">&lt;-</span> <span class="va">lapis</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lapis</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># TODO: user-friendly parsing</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lapis</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lapis</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">provoc</span><span class="fu">:::</span><span class="fu"><a href="../reference/parse_mutation.html">parse_mutation</a></span><span class="op">(</span></span>
<span>        type <span class="op">=</span> <span class="st">"~"</span>,</span>
<span>        pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>        alt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_lineage_defs.html">plot_lineage_defs</a></span><span class="op">(</span><span class="va">lapis</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/get-lapis-1.png" width="700"></p>
<p>This looks a little different from before. There are 4 mutations that
seem to be in every lineage, which doesn’t give the de-mixing algorithm
any information about which lineages are present. B.1.1.7 seems to have
a lot of unique mutations, which might make the results more accurate
(if the mutations truly are unique to B.1.1.7).</p>
<p>Let’s quickly check how this compares to the barcodes:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_lineage_defs2.html">plot_lineage_defs2</a></span><span class="op">(</span><span class="va">bar</span>, <span class="va">astro</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>There are a couple mutations in astro but not bar, and a couple in
bar but not astro. This is to be expected. However, those red bars in
the shared mutation section are noteworthy! They are the same four
mutations that we saw in the plot before, which told us that the
de-mixing algorithm will have a hard time differentiating the lineages.
Let’s check the results.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_lapis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/provoc.html">provoc</a></span><span class="op">(</span><span class="va">count</span> <span class="op">/</span> <span class="va">coverage</span> <span class="op">~</span> <span class="va">.</span>,</span>
<span>    lineage_defs <span class="op">=</span> <span class="va">lapis</span>,</span>
<span>    data <span class="op">=</span> <span class="va">b3</span>, by <span class="op">=</span> <span class="st">"sra"</span><span class="op">)</span></span>
<span><span class="va">res_lapis</span><span class="op">$</span><span class="va">date</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lubridate.tidyverse.org/reference/ymd.html" class="external-link">ymd</a></span><span class="op">(</span><span class="va">res_lapis</span><span class="op">$</span><span class="va">date</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gg_lapis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">res_lapis</span>, date_col <span class="op">=</span> <span class="st">"date"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"LAPIS (CoV-Spectrum)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gg_astro</span> <span class="op">/</span> <span class="va">gg_bar</span> <span class="op">/</span> <span class="va">gg_lapis</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="op">~</span> <span class="va">lineage</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/gg-astro-bar-lapis-1.png" width="700"></p>
<p>Usher Barcodes and Cov-Spectrum generally agree, but what’s going on
with B.1.617.2??? It sort of looks like B.1.429 has taken some of the
weight from B.1.617.2, but this is worth looking into. For example, did
B.1.429 have mutations that are in B.1.617.2? From the comparison plot,
there appear to be a lot of mutations that differentiate B.1.429 and
B.1.617.2, so that can’t be it. Maybe it’s the coverage?</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_actual_defs.html">plot_actual_defs</a></span><span class="op">(</span><span class="va">res_lapis</span><span class="op">)</span></span></code></pre></div>
<p><img src="mutation_definitions_files/figure-html/lapis-coverage-1.png" width="700"></p>
<p>It doesn’t look like the mutations that differentiate B.1.617.2 and
B.1.429 had low coverage. The minimum of the average coverages across
samples<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;That is, for each mutation and lineage, I found the
average coverage across samples. The numbers on the plot below the
lineage names represent the minimum, median, and maximum of the
coverages shown in the plot.&lt;/p&gt;"><sup>1</sup></a>
isn’t low for the shared mutations.</p>
<p>I think the primary conclusion here is that the difference in
mutations between the barcodes and the lapis definitions is important
enough to the demixing algorithm to give the differences in results.
Perhaps the next plotting function I make should show something about
the frequencies of the mutations in each lineage.</p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Hopefully this vignette has successfully completed its two goals:</p>
<ol style="list-style-type: decimal">
<li>Demonstrate some of the functionality of <code>provoc</code> with
lineage definitions.</li>
<li>Demonstrate the importance of choosing good lineage
definitions!</li>
</ol>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-AndersenlabFreyja2024" class="csl-entry">
<span>“Andersen-Lab/<span>Freyja</span>.”</span> 2024. Andersen
Laboratory @ Scripps Research.
</div>
<div id="ref-baaijensLineageAbundanceEstimation2022" class="csl-entry">
Baaijens, Jasmijn A., Alessandro Zulli, Isabel M. Ott, Ioanna Nika, Mart
J. van der Lugt, Mary E. Petrone, Tara Alpert, et al. 2022.
<span>“Lineage Abundance Estimation for <span>SARS-CoV-2</span> in
Wastewater Using Transcriptome Quantification Techniques.”</span>
<em>Genome Biology</em> 23 (1): 236. <a href="https://doi.org/10.1186/s13059-022-02805-9" class="external-link">https://doi.org/10.1186/s13059-022-02805-9</a>.
</div>
<div id="ref-chenCoVSpectrumAnalysisGlobally2022a" class="csl-entry">
Chen, Chaoran, Sarah Nadeau, Michael Yared, Philippe Voinov, Ning Xie,
Cornelius Roemer, and Tanja Stadler. 2022.
<span>“<span>CoV-Spectrum</span>: Analysis of Globally Shared
<span>SARS-CoV-2</span> Data to Identify and Characterize New
Variants.”</span> <em>Bioinformatics</em> 38 (6): 1735–37. <a href="https://doi.org/10.1093/bioinformatics/btab856" class="external-link">https://doi.org/10.1093/bioinformatics/btab856</a>.
</div>
<div id="ref-CovlineagesConstellations2023" class="csl-entry">
<span>“Cov-Lineages/Constellations.”</span> 2023. CoV-lineages.
</div>
<div id="ref-karthikeyanWastewaterSequencingReveals2022" class="csl-entry">
Karthikeyan, Smruthi, Joshua I. Levy, Peter De Hoff, Greg Humphrey,
Amanda Birmingham, Kristen Jepsen, Sawyer Farmer, et al. 2022.
<span>“Wastewater Sequencing Reveals Early Cryptic
<span>SARS-CoV-2</span> Variant Transmission.”</span> <em>Nature</em>,
July, 1–4. <a href="https://doi.org/10.1038/s41586-022-05049-6" class="external-link">https://doi.org/10.1038/s41586-022-05049-6</a>.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Devan Becker.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
